### PR TITLE
Feature google drive

### LINKED ISSUE #2709: [FEATURE] Google Drive auto-backups
### Please confirm the following:

- [X] I've checked [the current issues](https://github.com/Ivy-Apps/ivy-wallet/issues) for duplicate issues.
- [X] I've requested a single (only one) feature/change in this issue. It complies with the One Request Per GitHub Issue (ORPGI) rule.
- [X] My issue is well-defined and describes how it should be implemented from UI/UX perspective.

### What do you want to be added or improved?

As a user I want to be able to automatically backup my Ivy Wallet data in my personal Google Drive.

### Why do you need it?

- Ivy Wallet needs a free cloud backup feature
- Ivy Wallet needs automatic bacupsbackups

### How do you imagine it?

- The user logs in their Google account
- The user gives Google Drive permissions
- Ivy Wallet starts automatically attempting to backup a JSON zip every 12/24 hours

### PR DESCRIPTION
**Pull request by mohammednawas8**
## Pull Request (PR) Checklist
Please check if your pull request fulfills the following requirements:
- [X] The PR is submitted to the `main` branch.
- [X] I've read the [Contribution Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/CONTRIBUTING.md) and my PR doesn't break the "Contributing Rules".
- [X] I've read the [Architecture Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/Architecture.md).
- [X] I confirm that I've run the code locally and everything works as expected.
- [X] 🎬 I've attached a **screen recoding** of the changes. 

> Tip: drag & drop the video to the PR description.
https://github.com/Ivy-Apps/ivy-wallet/assets/78867217/5a972c68-ffbe-423d-b292-c20ce548f15a

## What's changed?

Describe with a few bullets **what's new:**
a Adding new option in the settings screen which is Google Drive backup
b Ability to link a google drive account to backup the user's data each 12 hours
c Ability to delete/change the drive account.

> 💡 Tip: Please, attach screenshots and screen recordings. It helps a lot!

## Risk Factors

**What may go wrong if we merge your PR?**

- a
- b
- c

**In what cases your code won't work?**

- a
- b
- c

## Does this PR closes any GitHub Issues?

Check **[Ivy Wallet Issues](https://github.com/Ivy-Apps/ivy-wallet/issues)**.

- Closes #2709 

> Replace `ISSUE_NUMBER` with your issue number (for example Closes #1234). If you've done that correctly, you'll see the issue title linked when previewing your PR description.

## Troubleshooting CI failures

If you see any of the PR checks failing (❌) go to [Actions](https://github.com/Ivy-Apps/ivy-wallet/actions) and find it there. Or simply click "Details" next to the failed check and explore the logs to see why it has failed.

### Detekt
[Detekt](https://detekt.dev/) is a static code analyzer for Kotlin that we use to enforce code readibility and good practices.

**To run Detekt locally:**
```
./gradlew detekt
```

If the Detekt errors are caused by a legacy code, you can suppress them using a basline.

**Detekt baseline** (not recommended)
```
./scripts/detektBaseline.sh
```

### Lint

We use the [standard Android Lint](https://developer.android.com/studio/write/lint) plus [Slack's compose-lints](https://slackhq.github.io/compose-lints/) as an addition to enforce proper Compose usage.

**To run Lint locally:**
```
./scripts/lint.sh
```

If the Lint errors are caused by a legacy code, you can suppress them using a basline.

**Lint baseline** (not recommended)
```
./scripts/lintBaseline.sh
```

### Unit tests

If this job is failing this means that your changes break an existing unit test. You must identify the failing tests and fix your code.

**To run the Unit tests locally:**
```
./gradlew testDebugUnitTest
```

### CODE REVIEW

File: .idea/codeStyles/codeStyleConfig.xml
Comment (ILIYANGERMANOV) [inline .idea/codeStyles/codeStyleConfig.xml:1]: Remove + add to `.gitignore`

File: .idea/shelf/Uncommitted_changes_before_Checkout_at_18_09_2023/shelved.patch
Comment (ILIYANGERMANOV) [inline .idea/shelf/Uncommitted_changes_before_Checkout_at_18_09_2023/shelved.patch:1]: Remove + add to `.gitignore`

File: app/src/main/java/com/ivy/wallet/IvyAndroidApp.kt [hunk 1]
```diff
@@ -31,4 +36,26 @@ class IvyAndroidApp : Application(), Configuration.Provider {
             Timber.plant(DebugTree())
         }
     }
-}
\ No newline at end of file
+}
+
+// This solves Could not instantiat worker issue
+class CustomWorkerFactory @Inject constructor(
```
Comment (ILIYANGERMANOV) [inline app/src/main/java/com/ivy/wallet/IvyAndroidApp.kt:42]: Is there a better solution to this? This seems like a hack that'll become tech debt in the future

File: config/detekt/baseline.yml [hunk 2]
```diff
@@ -2390,6 +2392,7 @@
     <ID>MagicNumber:BackupLogic.kt$BackupLogic$0.7</ID>
     <ID>MagicNumber:BackupLogic.kt$BackupLogic$0.8</ID>
     <ID>MagicNumber:BackupLogic.kt$BackupLogic$0.9</ID>
+    <ID>MagicNumber:BackupWorker.kt$BackupWorker$5</ID>
```
Comment (ILIYANGERMANOV) [inline config/detekt/baseline.yml:2395]: This is valid Detekt feedback. Can you revert the baseline and fix as many Detekt errors as possible? The ones in your module seems valid.

File: config/detekt/baseline.yml [hunk 3]
```diff
@@ -5567,6 +5588,12 @@
     <ID>NoUnusedImports:DefaultConfigurationAccessors.kt$gradle.kotlin.dsl.accessors._c3b5a8584877c7599cd6fe5cb3cf4f8d.DefaultConfigurationAccessors.kt</ID>
     <ID>NoUnusedImports:DetektConfigurationAccessors.kt$gradle.kotlin.dsl.accessors._6b1cdd1e881959619ea23cf7941079a9.DetektConfigurationAccessors.kt</ID>
     <ID>NoUnusedImports:DetektPluginsConfigurationAccessors.kt$gradle.kotlin.dsl.accessors._6b1cdd1e881959619ea23cf7941079a9.DetektPluginsConfigurationAccessors.kt</ID>
+    <ID>NoUnusedImports:DriveBackupConstants.kt$com.ivy.googledrive.backup.DriveBackupConstants.kt</ID>
```
Comment (ILIYANGERMANOV) [inline config/detekt/baseline.yml:5591]: Those are all valid. Please, revert the baseline

File: google-drive/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+plugins {
+    id("ivy.feature")
+    kotlin("kapt")
```
Comment (ILIYANGERMANOV) [inline google-drive/build.gradle.kts:3]: Why "kapt"? It makes the builds slow and we no longer need it.

File: google-drive/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+android {
+    namespace = "com.ivy.googledrive"
+}
+
+dependencies {
+    implementation(projects.ivyBase)
+    implementation(projects.ivyResources)
+    implementation(projects.tempLegacyCode)
+    api(libs.google.services.drive) {
```
Comment (ILIYANGERMANOV) [inline google-drive/build.gradle.kts:14]: "api" is banned because it leads to complexity and slow build times, use `implementation`

File: google-drive/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        exclude(group = "com.google.guava", module = "listenablefuture")
+    }
+    implementation(libs.google.http.gson)
+    implementation(libs.google.http.android) {
+        exclude(group = "com.google.guava", module = "listenablefuture")
+    }
+    implementation(libs.google.guava)
+    api(libs.google.playservices.auth)
```
Comment (ILIYANGERMANOV) [inline google-drive/build.gradle.kts:22]: same

File: google-drive/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+    implementation(libs.bundles.hilt)
+    api(libs.androidx.work)
```
Comment (ILIYANGERMANOV) [inline google-drive/build.gradle.kts:25]: same

File: google-drive/src/main/java/com/ivy/googledrive/backup/BackupWorker.kt [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        try {
+            val driveBackupRepository = getDriveBackRepository(appContext)
+            val jsonContent = backupLogic.generateJsonBackup()
+            val backupFile = File(appContext.filesDir, "${IVY_BACKUP_FILE_NAME}.json")
+            backupFile.writeText(jsonContent)
+            val fileId = driveBackupRepository.backupFile(
+                backupFile,
+                "${IVY_BACKUP_FILE_NAME}-${Date()}",
+                DriveMimeType.JSON,
+                5
```
Comment (ILIYANGERMANOV) [inline google-drive/src/main/java/com/ivy/googledrive/backup/BackupWorker.kt:43]: Magic Number, Detekt raised a valid error

File: google-drive/src/main/java/com/ivy/googledrive/backup/BackupWorker.kt [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            Timber.tag(TAG).e("Error while backing up the user's data")
+            e.printStackTrace()
+            return Result.failure()
+        }
+    }
+
+    // The reason of why i did not use hilt to inject this repo in the constructor is because i want to get the drive instance during
+    // The runtime, if we use hilt to inject this in the constructor the drive instance will be null when the user links his google account
+    // For the first time.
+    private fun getDriveBackRepository(context: Context): DriveBackupRepository {
```
Comment (ILIYANGERMANOV) [inline google-drive/src/main/java/com/ivy/googledrive/backup/BackupWorker.kt:62]: You can use `@Provides` + dagger.Lazy like `Lazy<DriveBackupRepository` - let's use Hilt for that

File: google-drive/src/main/java/com/ivy/googledrive/backup/DriveBackupConstants.kt [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package com.ivy.googledrive.backup
+
+import java.util.UUID
+
+object DriveBackupConstants {
+
+    internal const val IVY_FOLDER_NAME = "Ivy"
+    const val IVY_BACKUP_FILE_NAME = "Ivy-backup"
+    const val WORKER_TAG = "Ivy-Backup-Worker"
+
```
Comment (ILIYANGERMANOV) [inline google-drive/src/main/java/com/ivy/googledrive/backup/DriveBackupConstants.kt:10]: Needless blank lines, Detekt has caught that too

File: google-drive/src/main/java/com/ivy/googledrive/backup/DriveBackupRepository.kt [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+import com.ivy.googledrive.google_drive.DriveMimeType
+import java.io.File
+import kotlin.jvm.Throws
+
+interface DriveBackupRepository {
+    /**
+     * @param n is The capacity of the Backup folder,
+     * if the capacity is full then the oldest file will get deleted and get replaced by the new file.
+     * */
+    @Throws(Exception::class)
```
Comment (ILIYANGERMANOV) [inline google-drive/src/main/java/com/ivy/googledrive/backup/DriveBackupRepository.kt:13]: That's very bad! Throwing exceptions leads to unexpected crash. Please use Arrow's `Either` like we did in https://github.com/Ivy-Apps/ivy-wallet/blob/deprecated-ivywallet-rewrite-2223/drive/google-drive/src/main/java/com/ivy/drive/google_drive/GoogleDriveServiceImpl.kt

File: google-drive/src/main/java/com/ivy/googledrive/backup/DriveBackupRepository.kt [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    suspend fun backupFile(
+        file: File,
+        fileName: String,
+        mimeType: DriveMimeType,
+        n: Int
```
Comment (ILIYANGERMANOV) [inline google-drive/src/main/java/com/ivy/googledrive/backup/DriveBackupRepository.kt:18]: What is `n`? 😄

**Review by ILIYANGERMANOV — CHANGES_REQUESTED** [review]
Hey @mohammednawas8 thank you for your hard work! 🎉 Your code seems working but it we need to make some changes to make it mergeable.

You need to address:
- proper error handling using Arrow (https://github.com/Ivy-Apps/ivy-wallet/blob/deprecated-ivywallet-rewrite-2223/drive/google-drive/src/main/java/com/ivy/drive/google_drive/GoogleDriveServiceImpl.kt)
- Don't suppress Detekt errors in your module - fix them.
- Add Unit tests using Mockk for all classes
- Ideally, it should look like the implementation which we did last year:
https://github.com/Ivy-Apps/ivy-wallet/tree/deprecated-ivywallet-rewrite-2223/drive/google-drive/src/main/java/com/ivy/drive/google_drive

I've pasted in the issue for reference because it features solid error handling + good type-safety.

No rush, man. If you're near the situation in Israel, please stay safe. Hope this ends soon 🙏

**Comment by ILIYANGERMANOV** [comment]
Hey @mohammednawas8 we also need to indicate whether Google Drive backups are enabled in Settings or not. Maybe a Switch? Or... some UI telling the user whether backups are enabled.

**Comment by mohammednawas8** [comment]
Did you watch the attached video
https://github-production-user-asset-6210df.s3.amazonaws.com/78867217/274605831-5a972c68-ffbe-423d-b292-c20ce548f15a.mp4

as you can notice if the user has already linked his drive account with the app and tried to click on the google drive backup option again, a dialog will show up that shows the back up email and the ability to delete the account.

**Comment by ILIYANGERMANOV** [comment]
> Did you watch the attached video https://github-production-user-asset-6210df.s3.amazonaws.com/78867217/274605831-5a972c68-ffbe-423d-b292-c20ce548f15a.mp4
> 
> as you can notice if the user has already linked his drive account with the app and tried to click on the google drive backup option again, a dialog will show up that shows the back up email and the ability to delete the account.

the user must be able to notice that w/o having to click. Like having different UI states on the button. If enabled for example a green pill saying "Enabled"

**Comment by mohammednawas8** [comment]
okay i will work on it

**Comment by mohammednawas8** [comment]
I added a green text in the same option
https://github.com/Ivy-Apps/ivy-wallet/assets/78867217/695408f3-daba-4e53-a46e-8a2b0d97191d

**Comment by ILIYANGERMANOV** [comment]
> I added a green text in the same option https://github.com/Ivy-Apps/ivy-wallet/assets/78867217/695408f3-daba-4e53-a46e-8a2b0d97191d

Screenshot_20231012_222438_Chrome

Remove the ✔️  and use only "Enabled" with 1dp border radius green

**Comment by mohammednawas8** [comment]
Like this
enable flag

**Comment by ILIYANGERMANOV** [comment]
> Like this enable flag

It's a perfect! Just do
- vertical padding 4.dp or 8.dp
- horizontal 16.dp

File: google-drive/src/main/java/com/ivy/googledrive/backup/DriveBackupRepository.kt (hunk 12 above)
Comment (mohammednawas8) [reply google-drive/src/main/java/com/ivy/googledrive/backup/DriveBackupRepository.kt:18]: it is the backup folder capacity, you mentioned it in the requirements
* When N is reached the oldest backup is deleted.

File: google-drive/src/main/java/com/ivy/googledrive/backup/DriveBackupRepository.kt (hunk 11 above)
Comment (mohammednawas8) [reply google-drive/src/main/java/com/ivy/googledrive/backup/DriveBackupRepository.kt:13]: I catch the exceptions using try and catch

Comment (ILIYANGERMANOV) [reply google-drive/src/main/java/com/ivy/googledrive/backup/DriveBackupRepository.kt:13]: Hey @mohammednawas8 that works but the problem is that Exceptions are prone to errors and runtime crashes. We work with typed errors:
https://arrow-kt.io/learn/typed-errors/working-with-typed-errors/

**TL;DR;**

- If a function is expected to fail, e.g. Network request (it happens often) we model its return type explicitly with `Either`
- We throw Exceptions only for exceptional cases - situations which we don't expect to happen. For example, save to DB failing because the device is out of storage or some IO error.

Does that make sense?

Comment (ILIYANGERMANOV) [reply google-drive/src/main/java/com/ivy/googledrive/backup/DriveBackupRepository.kt:13]: This can provide more context:
https://www.linkedin.com/posts/iliyan-germanov-3963b5b9_working-with-typed-errors-arrow-activity-7115666238504103937-d5df?utm_source=share&utm_medium=member_android
